diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1b9ce82..725cac8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -458,7 +458,7 @@ set_target_properties(portaudio PROPERTIES
   PUBLIC_HEADER "${PORTAUDIO_PUBLIC_HEADERS}"
   MACOSX_FRAMEWORK_IDENTIFIER com.portaudio
   FRAMEWORK_VERSION A
-  WINDOWS_EXPORT_ALL_SYMBOLS TRUE
+  WINDOWS_EXPORT_ALL_SYMBOLS FALSE
   VERSION ${PROJECT_VERSION}
   SOVERSION 2
 )
